torch/cuda/__init__.py

```diff
@@ -317,41 +326,118 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
         raise AssertionError("Invalid peer device id")
     return torch._C._cuda_canDeviceAccessPeer(device, peer_device)
 
-
-@contextlib.contextmanager
-def stream(stream):
+class StreamContext(object):
     r"""Context-manager that selects a given stream.
-
     All CUDA kernels queued within its context will be enqueued on a selected
     stream.
-
-    Args:
-        stream (Stream): selected stream. This manager is a no-op if it's
+    Arguments:
+        StreamContext (Stream): selected stream. This manager is a no-op if it's
             ``None``.
-
     .. note:: Streams are per-device. If the selected stream is not on the
-        current device, this function will also change the current device to
-        match the stream.
+    current device, this function will also change the current device to
+    match the stream.
     """
-    if stream is None:
-        yield
-        return
-    src_prev_stream = current_stream()
-
-    if src_prev_stream.device != stream.device:
-        # The given stream is on a different device; have to restore the
-        # current_stream on that device on exit as well
-        with device(stream.device):
-            dst_prev_stream = current_stream()
-
-    torch._C._cuda_setStream(stream._cdata)
-    try:
-        yield
-    finally:
-        if src_prev_stream.device != stream.device:
-            torch._C._cuda_setStream(dst_prev_stream._cdata)
-        torch._C._cuda_setStream(src_prev_stream._cdata)
+    cur_stream : Optional['torch.classes.cuda.Stream']  # type: ignore
+    cur_stream_device: Optional[_device]
+
+    def __init__(self, stream: Optional['torch.classes.cuda.Stream']):  # type: ignore
+        self.idx = -1
+        self.stream = stream
+        self.prev_stream_device_index = -1
+        self.cur_stream_device_index = -1
```

Comment:
Given that you can query the device and device index of a stream, do we need these as separate pieces of state in the context manager object?

Reply:
We can. Both the indices are referenced in `__enter__` and `__exit__` of the context manager. In the current implementation, the two index are initialized in the `__enter__` and then read in the `__exit__`. If we do not maintain the state, then we would have to  duplicate the same code which is in `__enter__`. This will again bloat up the context manager.

Reply:
I do not understand. If we can query the device and device index of a stream, why is it not sufficient to store only the stream?

Reply:
Can you add your reply as a comment? I think you edited my comment instead and it's harder for me to understand now.

Reply:
Oops, yes I realized that and removed the reply. Sorry about that.
Yes, Looks like only `operator != ` is not supported. So, yes that's a great idea. We can then eliminate the index part. 
Also, like you mentioned, if this PR  #50670 will be merged soon, then I can just retain the previous eager code `if src_prev_stream.device != stream.device`